File: repos/openjarvis/src/openjarvis/security/capabilities.py
"""RBAC capability system — fine-grained permission model for tool dispatch."""

from __future__ import annotations

import fnmatch
import json
import logging
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Dict, List, Optional

logger = logging.getLogger(__name__)


class Capability(str, Enum):
    """Fine-grained capability labels."""
    FILE_READ = "file:read"
    FILE_WRITE = "file:write"
    NETWORK_FETCH = "network:fetch"
    CODE_EXECUTE = "code:execute"
    MEMORY_READ = "memory:read"
    MEMORY_WRITE = "memory:write"
    CHANNEL_SEND = "channel:send"
    TOOL_INVOKE = "tool:invoke"
    SCHEDULE_CREATE = "schedule:create"
    SYSTEM_ADMIN = "system:admin"


@dataclass(slots=True)
class CapabilityGrant:
    """A single capability grant for an agent."""
    capability: str              # Capability value or glob pattern
    pattern: str = "*"           # resource glob pattern


@dataclass(slots=True)
class AgentPolicy:
    """Policy for a specific agent."""
    agent_id: str
    grants: List[CapabilityGrant] = field(default_factory=list)
    deny: List[str] = field(default_factory=list)  # explicit denials


class CapabilityPolicy:
    """RBAC capability policy for tool dispatch.

    Checks whether an agent has the required capability to invoke a tool.
    Policy can be loaded from a JSON file or configured programmatically.

    Default policy: if no explicit policy exists for an agent, all
    capabilities are granted (open by default). Set ``default_deny=True``
    to flip to deny-by-default.
    """

    def __init__(
        self,
        *,
        policy_path: Optional[str] = None,
        default_deny: bool = False,
    ) -> None:
        self._policies: Dict[str, AgentPolicy] = {}
        self._default_deny = default_deny

        from openjarvis._rust_bridge import get_rust_module
        _rust = get_rust_module()
        self._rust_impl = _rust.CapabilityPolicy(default_deny=default_deny)

        if policy_path:
            self._load_file(Path(policy_path))

    def grant(self, agent_id: str, capability: str, pattern: str = "*") -> None:
        """Grant a capability to an agent."""
        policy = self._policies.setdefault(
            agent_id, AgentPolicy(agent_id=agent_id),
        )
        policy.grants.append(CapabilityGrant(capability=capability, pattern=pattern))
        self._rust_impl.grant(agent_id, capability, pattern)

    def deny(self, agent_id: str, capability: str) -> None:
        """Explicitly deny a capability to an agent."""
        policy = self._policies.setdefault(
            agent_id, AgentPolicy(agent_id=agent_id),
        )
        policy.deny.append(capability)
        self._rust_impl.deny(agent_id, capability)

    def check(self, agent_id: str, capability: str, resource: str = "") -> bool:
        """Check whether *agent_id* has *capability* for *resource*.

        Returns True if allowed, False if denied.
        """
        return self._rust_impl.check(agent_id, capability, resource)

    def _check_python(self, agent_id: str, capability: str, resource: str = "") -> bool:
        """Legacy Python check — kept for reference only."""
        policy = self._policies.get(agent_id)
        if policy is None:
            # No explicit policy — use default
            return not self._default_deny

        # Explicit denials take precedence
        for denied in policy.deny:
            if fnmatch.fnmatch(capability, denied):
                return False

        # Check grants
        for grant in policy.grants:
            if fnmatch.fnmatch(capability, grant.capability):
                if resource and grant.pattern != "*":
                    if fnmatch.fnmatch(resource, grant.pattern):
                        return True
                else:
                    return True

        # No matching grant found
        return not self._default_deny

    def list_grants(self, agent_id: str) -> List[CapabilityGrant]:
        """List all grants for an agent."""
        policy = self._policies.get(agent_id)
        return list(policy.grants) if policy else []

    def list_agents(self) -> List[str]:
        """List all agents with explicit policies."""
        return list(self._policies.keys())

    def _load_file(self, path: Path) -> None:
        """Load policy from a JSON file."""
        if not path.exists():
            return
        try:
            data = json.loads(path.read_text())
            for agent_data in data.get("agents", []):
                agent_id = agent_data["agent_id"]
                for grant_data in agent_data.get("grants", []):
                    self.grant(
                        agent_id,
                        grant_data["capability"],
                        grant_data.get("pattern", "*"),
                    )
                for denied in agent_data.get("deny", []):
                    self.deny(agent_id, denied)
        except (json.JSONDecodeError, KeyError, TypeError) as exc:
            logger.warning("Failed to parse capability policy: %s", exc)

    def save(self, path: Path) -> None:
        """Save policy to a JSON file."""
        agents = []
        for agent_id, policy in self._policies.items():
            agents.append({
                "agent_id": agent_id,
                "grants": [
                    {"capability": g.capability, "pattern": g.pattern}
                    for g in policy.grants
                ],
                "deny": policy.deny,
            })
        path.write_text(json.dumps({"agents": agents}, indent=2))


# Default capability requirements for built-in tools
DEFAULT_TOOL_CAPABILITIES: Dict[str, List[str]] = {
    "file_read": [Capability.FILE_READ],
    "web_search": [Capability.NETWORK_FETCH],
    "code_interpreter": [Capability.CODE_EXECUTE],
    "memory_store": [Capability.MEMORY_WRITE],
    "memory_retrieve": [Capability.MEMORY_READ],
    "memory_search": [Capability.MEMORY_READ],
    "memory_index": [Capability.MEMORY_WRITE],
    "schedule_task": [Capability.SCHEDULE_CREATE],
    "channel_send": [Capability.CHANNEL_SEND],
}


__all__ = [
    "AgentPolicy",
    "Capability",
    "CapabilityGrant",
    "CapabilityPolicy",
    "DEFAULT_TOOL_CAPABILITIES",
]

File: repos/openjarvis/rust/crates/openjarvis-security/src/capabilities.rs
//! RBAC capability system — fine-grained permission model for tool dispatch.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Capability {
    #[serde(rename = "file:read")]
    FileRead,
    #[serde(rename = "file:write")]
    FileWrite,
    #[serde(rename = "network:fetch")]
    NetworkFetch,
    #[serde(rename = "code:execute")]
    CodeExecute,
    #[serde(rename = "memory:read")]
    MemoryRead,
    #[serde(rename = "memory:write")]
    MemoryWrite,
    #[serde(rename = "channel:send")]
    ChannelSend,
    #[serde(rename = "tool:invoke")]
    ToolInvoke,
    #[serde(rename = "schedule:create")]
    ScheduleCreate,
    #[serde(rename = "system:admin")]
    SystemAdmin,
}

impl Capability {
    pub fn as_str(&self) -> &'static str {
        match self {
            Capability::FileRead => "file:read",
            Capability::FileWrite => "file:write",
            Capability::NetworkFetch => "network:fetch",
            Capability::CodeExecute => "code:execute",
            Capability::MemoryRead => "memory:read",
            Capability::MemoryWrite => "memory:write",
            Capability::ChannelSend => "channel:send",
            Capability::ToolInvoke => "tool:invoke",
            Capability::ScheduleCreate => "schedule:create",
            Capability::SystemAdmin => "system:admin",
        }
    }
}

#[derive(Debug, Clone)]
pub struct CapabilityGrant {
    pub capability: String,
    pub pattern: String,
}

#[derive(Debug, Clone)]
struct AgentPolicy {
    grants: Vec<CapabilityGrant>,
    deny: Vec<String>,
}

/// RBAC capability policy for tool dispatch.
///
/// Default policy: if no explicit policy exists for an agent, all
/// capabilities are granted. Set `default_deny` to flip.
pub struct CapabilityPolicy {
    policies: HashMap<String, AgentPolicy>,
    default_deny: bool,
}

impl CapabilityPolicy {
    pub fn new(default_deny: bool) -> Self {
        Self {
            policies: HashMap::new(),
            default_deny,
        }
    }

    pub fn grant(&mut self, agent_id: &str, capability: &str, pattern: &str) {
        let policy = self.policies.entry(agent_id.to_string()).or_insert_with(|| {
            AgentPolicy {
                grants: Vec::new(),
                deny: Vec::new(),
            }
        });
        policy.grants.push(CapabilityGrant {
            capability: capability.to_string(),
            pattern: pattern.to_string(),
        });
    }

    pub fn deny(&mut self, agent_id: &str, capability: &str) {
        let policy = self.policies.entry(agent_id.to_string()).or_insert_with(|| {
            AgentPolicy {
                grants: Vec::new(),
                deny: Vec::new(),
            }
        });
        policy.deny.push(capability.to_string());
    }

    pub fn check(&self, agent_id: &str, capability: &str, resource: &str) -> bool {
        let policy = match self.policies.get(agent_id) {
            Some(p) => p,
            None => return !self.default_deny,
        };

        for denied in &policy.deny {
            if glob_match(denied, capability) {
                return false;
            }
        }

        for grant in &policy.grants {
            if glob_match(&grant.capability, capability) {
                if !resource.is_empty() && grant.pattern != "*" {
                    if glob_match(&grant.pattern, resource) {
                        return true;
                    }
                } else {
                    return true;
                }
            }
        }

        !self.default_deny
    }

    pub fn list_agents(&self) -> Vec<String> {
        self.policies.keys().cloned().collect()
    }

    pub fn load_json(&mut self, json_str: &str) -> Result<(), serde_json::Error> {
        let data: serde_json::Value = serde_json::from_str(json_str)?;
        if let Some(agents) = data["agents"].as_array() {
            for agent_data in agents {
                let agent_id = agent_data["agent_id"].as_str().unwrap_or("");
                if agent_id.is_empty() {
                    continue;
                }
                if let Some(grants) = agent_data["grants"].as_array() {
                    for g in grants {
                        let cap = g["capability"].as_str().unwrap_or("");
                        let pat = g["pattern"].as_str().unwrap_or("*");
                        self.grant(agent_id, cap, pat);
                    }
                }
                if let Some(deny_list) = agent_data["deny"].as_array() {
                    for d in deny_list {
                        if let Some(cap) = d.as_str() {
                            self.deny(agent_id, cap);
                        }
                    }
                }
            }
        }
        Ok(())
    }
}

impl Default for CapabilityPolicy {
    fn default() -> Self {
        Self::new(false)
    }
}

fn glob_match(pattern: &str, text: &str) -> bool {
    if pattern == "*" {
        return true;
    }
    if pattern == text {
        return true;
    }
    let parts: Vec<&str> = pattern.split('*').collect();
    if parts.len() == 1 {
        return pattern == text;
    }

    let mut pos = 0;
    for (i, part) in parts.iter().enumerate() {
        if part.is_empty() {
            continue;
        }
        if let Some(found) = text[pos..].find(part) {
            if i == 0 && found != 0 {
                return false;
            }
            pos += found + part.len();
        } else {
            return false;
        }
    }
    if let Some(last) = parts.last() {
        if !last.is_empty() && !text.ends_with(last) {
            return false;
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_allow() {
        let policy = CapabilityPolicy::new(false);
        assert!(policy.check("agent1", "file:read", ""));
    }

    #[test]
    fn test_default_deny() {
        let policy = CapabilityPolicy::new(true);
        assert!(!policy.check("agent1", "file:read", ""));
    }

    #[test]
    fn test_explicit_grant() {
        let mut policy = CapabilityPolicy::new(true);
        policy.grant("agent1", "file:read", "*");
        assert!(policy.check("agent1", "file:read", ""));
        assert!(!policy.check("agent1", "file:write", ""));
    }

    #[test]
    fn test_explicit_deny_overrides_grant() {
        let mut policy = CapabilityPolicy::new(false);
        policy.grant("agent1", "file:*", "*");
        policy.deny("agent1", "file:write");
        assert!(policy.check("agent1", "file:read", ""));
        assert!(!policy.check("agent1", "file:write", ""));
    }

    #[test]
    fn test_glob_match() {
        assert!(glob_match("*", "anything"));
        assert!(glob_match("file:*", "file:read"));
        assert!(!glob_match("file:read", "file:write"));
        assert!(glob_match("*.txt", "doc.txt"));
    }
}
